### ISSUE TITLE
Page stops loading

### ISSUE THREAD

**Issue by Saiberion** [sergey-dryabzhinsky/qt-webkit-kiosk#14]
It can happen that loading hangs when changing webpages. Screen stays blank loading bar does not change.

**Comment by sergey-dryabzhinsky** [comment]
Page load timeout? Network errors?

What reaction is appropriate will be?
Show page-warning and try to reload after Nth seconds?

**Comment by Saiberion** [comment]
Network should not be the issue because the websize is served by the device itself.

I use qt-webkit-kiosk on an iMX6 controller and switching between local webpages may stop loading after several page loads (not easy to reproduce). That means progressbar usually stops at 50% and clicking anywhere causes everything to go white (like window is not responding in the windows world). Reloading helps but in my current setup I have no way to trigger a reload. At the moment I try to find out what causes the stop of loading. As of now it maybe Ajax related it seems like loading data is stuck in a loop.

Best would be to find what exactly causes the haning and fix it.

**Comment by sergey-dryabzhinsky** [comment]
Does any page have iframe in it?

**Comment by BluTiGeS** [comment]
I am also seeing that if the WebKit runs for some days (internal time update javascript called in interval).
, that the page does not react anymore. Have you also seen this?

**Comment by Saiberion** [comment]
I have seen this. In this case webkit-kiosk occupied all the RAM of the controller. But that's a completely different problem and seems to be related to JavaScript garbage collection and/or leaking JavaScript code.
My workaround for the hanging page load: I changed to a single site solution, only one page load when browser starts and first time loading always worked.
Workaround for high memory consumption (until I have to to analyse the reason): cron job to restart the browser once a week.

**Comment by sergey-dryabzhinsky** [comment]
1. I'll try to detect iframes and not fakeload them into main webview
2. Memory leak is in code with fakeload webview. Will try to fix it too.

**Comment by sergey-dryabzhinsky** [comment]
Iframes not fakeloading. That was something else.

**Comment by sergey-dryabzhinsky** [comment]
Try code from latest master branch.

**Comment by BluTiGeS** [comment]
Me or Saiberion?

Is the memory Leak also happening on default page loads? Where many page changes are done?

**Comment by sergey-dryabzhinsky** [comment]
Both of you I think.

Memory leak happened on "new window" loads. If link opens with "target=_blank".
If leak or freeze still happening then it's somewhere deep inside of QtWebKit.

Maybe QtWebEngine will work better. Maybe.

**Comment by sergey-dryabzhinsky** [comment]
And what version of Qt you are using?

**Comment by BluTiGeS** [comment]
I am using Qt 4.8.7, because no upgrade possible to 5.x possible with used platform.

Okay I do not open multiple New pages. I am just redirecting with JavaScript locaton.href  and ajax loadPage, so it shall be all happening in same window.

Will the performance changes you made bring a big improvement?

**Comment by sergey-dryabzhinsky** [comment]
I don't think so. No major performance improvement.

But reload page with JS can trigger "fakeload".
So there may be memory leak with old version of QWK.

**Comment by BluTiGeS** [comment]
I will give the code now a long term try.

Thanks Sergey

Regards

**Comment by BluTiGeS** [comment]
Hi all,

I have some updates here, pages are still stopping loading.

I didsome more debugging and added JS window.error function to print into my logs.

As a result I get, the  qt-webkit-kiosk still reacts on the killall -USR1 Signals but Javascript is not evaluated anymore. 

Simple HMTL pages are working without any scripts.
**TypeError: 'undefined' is not a function in undefined on line 0** 

But since it is at line 0 it must be something before page load, do you have any idea? It seems it occurs sporadic, sometimes running 1 day but sometimes several weeks.




Here they are seeing the same problem:
https://github.com/thoughtbot/capybara-webkit/issues/732

**Comment by sergey-dryabzhinsky** [comment]
So the simpe test would be a page with JS which reloads it every second.
- Do you have massive pages with big JS/CSS?
- Pages are static files or dynamic site content?

Need to simulate this issue correctly.

**Comment by BluTiGeS** [comment]
The page is quite large it has jquery+jquery ui +own objects included so yes the page is big.

The page is also running a interval which updates the time every second and a second interval which triggers info access every 10 seconds.

But in the state there it stops working the minmal page with just a location.href = index.php does not work anymore.

after a restart of qt-webkit-kiosk it is running flawless.


Also with the Singals to the qt-webkit-kiosk it does not change the behaviour still in a lockup state.

I am also wondering why a meta redirect did not work.....

**Comment by sergey-dryabzhinsky** [comment]
Did you check qt-webkit-kiosk process memory size? May be it leaking somewhere.
And could you try start QWK with disk cache disabled in config.

**Comment by BluTiGeS** [comment]
I checked it but it was in a good shape 100M Virtual and 37M RES, from top command in embedded LINUX
I think I ahve disk cache disabled in Config but I can check tomorrow. But disk space was also enough free.

**Comment by BluTiGeS** [comment]
I checked the config an caching is disabled:

**Comment by sergey-dryabzhinsky** [comment]
It seems fixed in Qt5 version of QtWebKit.

> I am using Qt 4.8.7, because no upgrade possible to 5.x possible with used platform.

Why not possible? What platform are you using?

Did you tryed fix from http://stackoverflow.com/questions/5605588/how-to-use-requestanimationframe/18070493#18070493 ? Add code for ```window.requestAnimationFrame``` to html>head>script section.

**Comment by BluTiGeS** [comment]
Why shall adding reimplementaion of  a requestAnimationFrame solve this?
We are not using this requestAnimationFrame  in our code so far, or am I overseeing something here?

Maybe it is possible but a lot of effort, since  we are running on a old toolchain.

**Comment by sergey-dryabzhinsky** [comment]
It it possible not reload page but call some JS function on that page to do some job?
I can check USR1/USR2 handlers URI - if it's URI - reload page, if string - evaluate JS code.

I'm just thinking that some reload request may cause lockdown javascript engine.
May be it not stopped before second reload requested, runnig first request.
Like intersection.

**Comment by BluTiGeS** [comment]
if it is in this state you can call some basic javascript like alert and so on.


Also what I am seeing is that the timer update stops working on the systems.
Is there a way to run the webkit kiosk in a monotonic way? So that it is independent from the system clock? otherwise if the system clock changes to time in the past the intervals are out of scope at this time.

**Comment by BluTiGeS** [comment]
We are starting to do a upgrade to qt 5.x but we mentioned that webkit is dropped int qt 5.6 so it is not available anymore.

Does the qt-webkit-kiosk work with 5.6 and further versions?

**Comment by sergey-dryabzhinsky** [comment]
Not yeat. I started porting to QtWebEngine, but it take time.

**Comment by BluTiGeS** [comment]
Okay glad to hear.

Today we saw again that a signal cannot load the page. We have enabled the qDebug build and saw that the log entries show it is hung up at Loading of the page with 65% and did not continue.

Any idea why it does not finish the loading progress of the QWebview load?

Regards

**Comment by BluTiGeS** [comment]
Another interesting thing we found  out if you set the time for 3 min into the past and set it to current time again and redo this some times the javascript window.location.href did not get executed.

So maybe all is somehow related to the sysclock, but our embedded systems supports sysconf(_SC_MONOTONIC_CLOCK)


And as far as I know the timers in Qt shall use the monotonic clock if available. Do you see any reason why the setIntervals etc in the qt-webkit-kiosk get out of sync during a system time change?

**Comment by BluTiGeS** [comment]
Now again at same issue, page stops loading at 65 % also a triggered view->reload() did not help here, only a application restart solves the issue. Any suggestion why the load does not complete? In the logs i cannot see any hint

**Comment by BluTiGeS** [comment]
Today I found maybe a hint in the logs, since I have added the Network Error Reply Error String;

Loading progress:  100 
Finish loading...
Try to activate window...
checkPageStoppedLoading...
**handleNetworkReply ERROR: "Operation canceled"** 
Start loading...
Loading progress:  10 
...
Loading progress:  100

HANG STATE


I have seen this Operation Canceled in all Logs there the Hang occurs.

Do you know any reason why the qt-webkit-kiosk is not in a correct state anymore if this error occurs?
Any solution?

**Comment by Hofmann-A** [comment]
I got the same behavior. 

Every time the LAN cable ist disconnected (Link is Down) or with an "ifconfig eth0 down" webkit stops loading pages, even pages are delivered by Local Loopback.

Javascript is still working and reconnecting LAN also brings the system back to life.

I cant see why this ist connected.

**Comment by sergey-dryabzhinsky** [comment]
So, if there is network error, QWK still tries to do `fihishLoading` action.
And cripples itself.

I'll add more debuging output.
And try to avoid further actions if there is some network error.
And try to output error into window.

**Comment by sergey-dryabzhinsky** [comment]
Could you try latest code from `master`?

**Comment by Hofmann-A** [comment]
Did an upgrade from 1.99.3 to master.

At first did not compile due to missing include for <QLabel> and <QHBoxLayout> in mainwindow.h, got it running after fixing. (using QT5.6 LTS)

After disconnecting Ethernet following error is shown: 

> ; Network error: Network access is disabled.; Page not fully loaded

Im using lighttpd and php-fpm to deliver the page locally, so no network should be necessary. Any bug-fix or hint if I made another error would be great.

**Comment by sergey-dryabzhinsky** [comment]
Rewrited signal handling from WebView to page->mainFrame signals.
It reduce call of handlers alot.
Now QWK messes with main page DOM only once.
`loadProgress` still called many times as sub-resources loading.
QWK should be more stable now.

Also:
- added network state monitoring. If network back - reload current page.
- fixed includes in `mainwindow.h`

Please try latest code from `master`.

**Comment by Hofmann-A** [comment]
Now it is working for about 5-7 seconds after disconnecting Ethernet. But then again

> ; Network error: Network access is disabled.

**Comment by sergey-dryabzhinsky** [comment]
Did your name-servers settings in `/etc/resolv.conf` pointing outside local network?

**Comment by sergey-dryabzhinsky** [comment]
Added forcing network accessible state if has loopback interface active.
Hope it works.
@Hofmann-A please try again with latest code from `master`.

**Comment by Hofmann-A** [comment]
Checked `/etc/resolv.conf` and removed google nameserver (used for GSM Server Connection). No change.

Tried latest code from `master`

`WebView::handleNetworkReply` is called:

> "Fri Jul 28 13:31:24 2017" handleNetworkReply ERROR: QNetworkReply::NetworkError(OperationCanceledError) = "Operation canceled"
"Fri Jul 28 13:31:25 2017" handleNetworkReply ERROR: QNetworkReply::NetworkError(UnknownNetworkError) = "Network access is disabled."

`MainWindow::networkStateChanged` ist not called

**Comment by sergey-dryabzhinsky** [comment]
Hmm... Strange.
What is your network connection? Ethernet or other one?
Do you have more than one network devices?

**Comment by Hofmann-A** [comment]
It is a standalone device with a display and GUI over qt-webkit-kiosk.
Ethernet and WLAN connections to local network are optional for extended operation (e.g. operate via mobile phone)

Example `ifconfig` with Link Up

> eth0      Link encap:Ethernet  HWaddr 74:E1:4A:30:46:D7
          inet addr:10.1.8.20  Bcast:10.1.8.255  Mask:255.255.255.0
          inet6 addr: fe80::76e1:4aff:fe30:46d7/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:821 errors:0 dropped:0 overruns:0 frame:0
          TX packets:525 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:523842 (511.5 KiB)  TX bytes:72766 (71.0 KiB)

> lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:2977 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2977 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1
          RX bytes:2724379 (2.5 MiB)  TX bytes:2724379 (2.5 MiB)

and for Link Down

> eth0      Link encap:Ethernet  HWaddr 74:E1:4A:30:46:D7
          inet addr:10.1.8.20  Bcast:10.1.8.255  Mask:255.255.255.0
          inet6 addr: fe80::76e1:4aff:fe30:46d7/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:951 errors:0 dropped:0 overruns:0 frame:0
          TX packets:575 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:538175 (525.5 KiB)  TX bytes:76778 (74.9 KiB)

> lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:2977 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2977 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1
          RX bytes:2724379 (2.5 MiB)  TX bytes:2724379 (2.5 MiB)

**Comment by sergey-dryabzhinsky** [comment]
Is your eth0 interface configured staticaly?
Seems like Qt can't figure out if interface not really gone down if link dropped.

**Comment by sergey-dryabzhinsky** [comment]
Add more network errors handling.
Try to restart NetworkSession if network not available but interfaces are up.

If there is serious error with network it may dive into infinite loop.
Need to detect routing / dns problems.

@Hofmann-A Please try code from latest `master`.

**Comment by Hofmann-A** [comment]
It is nearly working now! Few seconds after

> [  366.333111] fec 2188000.ethernet eth0: Link is Down

I get in the console:

> "Mon Jul 31 12:56:37 2017" handleNetworkReply ERROR: QNetworkReply::NetworkError(UnknownNetworkError) = "Network access is disabled."

And as message on the display 

> Network error: Network access is disabled.
Network access is disabled.; Page not fully loaded

But 5s later the message from display disappears and everything works fine again.

**Comment by sergey-dryabzhinsky** [comment]
It's reloaded after that type of error. And Qt network session restarted to reinit internal state of interfaces.
Probably need to add message about pending reload.

**Comment by Hofmann-A** [comment]
Tested for several hours, updated webpage then rebootet. Now I get a `Segmentation Fault` Error every time. Reverted webpage changes, Segmentation Fault stays.

Went back day by day. 
`Version 1.99.6 pre` is ok, 
`More network errors handling - if network not accessible...` is bad and generates warning

> QObject::connect: No such signal WebView::qwkNetworkError(NetworkError,QString)

drives me crazy...

**Comment by sergey-dryabzhinsky** [comment]
Please update to latest `master`
- removed `setMinimumWith` to `messagesBox` - seems like it kills `QHBoxLayout` and app

**Comment by Hofmann-A** [comment]
Yes, it is working. Thank you so much!

**Comment by sergey-dryabzhinsky** [comment]
Weird bug with `setMinimumWidth` =(

**Comment by sergey-dryabzhinsky** [comment]
I'll close this.
